File: backend/twilio_service.py
from twilio.rest import Client
import os
from typing import Dict, Any
import re

class TwilioService:
    def __init__(self):
        # Load Twilio credentials from environment
        self.account_sid = os.getenv('TWILIO_ACCOUNT_SID')
        self.auth_token = os.getenv('TWILIO_AUTH_TOKEN')
        self.phone_number = os.getenv('TWILIO_PHONE_NUMBER')
        self.verify_service_sid = os.getenv('TWILIO_VERIFY_SERVICE_SID')
        
        if not all([self.account_sid, self.auth_token, self.phone_number]):
            print("Warning: Twilio credentials not found in environment variables")
            self.client = None
        else:
            self.client = Client(self.account_sid, self.auth_token)
    
    def format_phone_number(self, phone_number: str) -> str:
        """Format phone number to E.164 format for India"""
        # Remove all non-digit characters
        cleaned = re.sub(r'\D', '', phone_number)
        
        # Add country code if not present (assuming India +91)
        if len(cleaned) == 10:
            return f'+91{cleaned}'
        elif len(cleaned) == 12 and cleaned.startswith('91'):
            return f'+{cleaned}'
        elif cleaned.startswith('+'):
            return phone_number  # Already formatted
        
        return phone_number
    
    def send_sms(self, to: str, message: str) -> Dict[str, Any]:
        """Send SMS using Twilio"""
        if not self.client:
            return {
                'success': False,
                'error': 'Twilio client not initialized. Check credentials.'
            }
        
        try:
            formatted_to = self.format_phone_number(to)
            
            message = self.client.messages.create(
                body=message,
                from_=self.phone_number,
                to=formatted_to
            )
            
            return {
                'success': True,
                'sid': message.sid,
                'status': message.status
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    def send_otp(self, phone_number: str) -> Dict[str, Any]:
        """Send OTP using Twilio Verify Service"""
        if not self.client or not self.verify_service_sid:
            return {
                'success': False,
                'error': 'Twilio Verify service not configured'
            }
        
        try:
            formatted_phone = self.format_phone_number(phone_number)
            
            verification = self.client.verify \
                .v2 \
                .services(self.verify_service_sid) \
                .verifications \
                .create(to=formatted_phone, channel='sms')
            
            return {
                'success': True,
                'status': verification.status,
                'to': verification.to
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    def verify_otp(self, phone_number: str, code: str) -> Dict[str, Any]:
        """Verify OTP using Twilio Verify Service"""
        if not self.client or not self.verify_service_sid:
            return {
                'success': False,
                'error': 'Twilio Verify service not configured'
            }
        
        try:
            formatted_phone = self.format_phone_number(phone_number)
            
            verification_check = self.client.verify \
                .v2 \
                .services(self.verify_service_sid) \
                .verification_checks \
                .create(to=formatted_phone, code=code)
            
            return {
                'success': verification_check.status == 'approved',
                'status': verification_check.status
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }

# Global instance
twilio_service = TwilioService()

File: src/services/twilioService.ts
// Note: For security reasons, Twilio operations should ideally be handled on the backend
// This is a client-side implementation for development purposes

export interface SMSRequest {
  to: string;
  message: string;
}

export interface OTPVerificationRequest {
  to: string;
  code: string;
}

class TwilioService {
  private baseUrl: string;

  constructor() {
    // Use backend API for secure Twilio operations
    this.baseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000';
  }

  // Send SMS using backend API
  async sendSMS(request: SMSRequest): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/send-sms`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          phone_number: request.to,
          message: request.message
        }),
      });

      if (response.ok) {
        const data = await response.json();
        if (data.success) {
          console.log('SMS sent successfully:', data.sid);
          return true;
        } else {
          console.error('Failed to send SMS:', data.error);
          return false;
        }
      } else {
        const error = await response.json();
        console.error('Failed to send SMS:', error);
        return false;
      }
    } catch (error) {
      console.error('Error sending SMS:', error);
      return false;
    }
  }

  // Send OTP using backend API
  async sendOTP(phoneNumber: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/send-otp`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          phone_number: phoneNumber
        }),
      });

      if (response.ok) {
        const data = await response.json();
        if (data.success) {
          console.log('OTP sent successfully:', data.status);
          return true;
        } else {
          console.error('Failed to send OTP:', data.error);
          return false;
        }
      } else {
        const error = await response.json();
        console.error('Failed to send OTP:', error);
        return false;
      }
    } catch (error) {
      console.error('Error sending OTP:', error);
      return false;
    }
  }

  // Verify OTP using backend API
  async verifyOTP(phoneNumber: string, code: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/verify-otp`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          phone_number: phoneNumber,
          code: code
        }),
      });

      if (response.ok) {
        const data = await response.json();
        if (data.success) {
          console.log('OTP verification successful:', data.status);
          return true;
        } else {
          console.error('Failed to verify OTP:', data.error);
          return false;
        }
      } else {
        const error = await response.json();
        console.error('Failed to verify OTP:', error);
        return false;
      }
    } catch (error) {
      console.error('Error verifying OTP:', error);
      return false;
    }
  }

  // Format phone number to E.164 format
  formatPhoneNumber(phoneNumber: string): string {
    // Remove all non-digit characters
    const cleaned = phoneNumber.replace(/\D/g, '');
    
    // Add country code if not present (assuming India +91)
    if (cleaned.length === 10) {
      return `+91${cleaned}`;
    } else if (cleaned.length === 12 && cleaned.startsWith('91')) {
      return `+${cleaned}`;
    }
    
    return phoneNumber;
  }
}

export const twilioService = new TwilioService();
export default TwilioService;